backend/LexBoxApi/Services/HgService.cs

```diff
@@ -96,36 +129,46 @@ public async Task DeleteRepo(string code)
     public async Task ResetRepo(string code)
     {
         string timestamp = FileUtils.ToTimestamp(DateTimeOffset.UtcNow);
+        await PrepareEmptyRepo(code, $"{timestamp}__empty");
         await SoftDeleteRepo(code, $"{timestamp}__reset");
         //we must init the repo as uploading a zip is optional
-        await InitRepo(code);
+        await MoveEmptyRepoIntoPlace(code, $"{timestamp}__empty");
     }
 
     public async Task FinishReset(string code, Stream zipFile)
     {
-        using var archive = new ZipArchive(zipFile, ZipArchiveMode.Read);
-        await DeleteRepo(code);
-        var repoPath = PrefixRepoFilePath(code);
-        var dir = Directory.CreateDirectory(repoPath);
-        archive.ExtractToDirectory(repoPath);
+        string timestamp = FileUtils.ToTimestamp(DateTimeOffset.UtcNow);
+        var tempRepoName = $"{code}__${timestamp}__upload";
+        var tempRepoPath = Path.Combine(_options.Value.RepoPath, DELETED_REPO_FOLDER, tempRepoName);
+        var tempRepo = Directory.CreateDirectory(tempRepoPath);
+        // TODO: Is Task.Run superfluous here? Or a good idea? Don't know the ins and outs of what happens before the first await in an async method in ASP.NET Core...
+        await Task.Run(() =>
+        {
+            using var archive = new ZipArchive(zipFile, ZipArchiveMode.Read);
+            archive.ExtractToDirectory(tempRepoPath);
+        });
 
-        var hgPath = Path.Join(repoPath, ".hg");
+        var hgPath = Path.Join(tempRepoPath, ".hg");
         if (!Directory.Exists(hgPath))
         {
-            var hgFolder = Directory.EnumerateDirectories(repoPath, ".hg", SearchOption.AllDirectories)
+            var hgFolder = Directory.EnumerateDirectories(tempRepoPath, ".hg", SearchOption.AllDirectories)
                 .FirstOrDefault();
             if (hgFolder is null)
             {
-                await DeleteRepo(code);
-                await InitRepo(code); // we don't want 404s
+                // Don't want to leave invalid .zip contents lying around as they may have been quite large
+                Directory.Delete(tempRepoPath, true);
                 //not sure if this is the best way to handle this, might need to catch it further up to expose the error properly to tus
                 throw ProjectResetException.ZipMissingHgFolder();
             }
             //found the .hg folder, move it to the correct location and continue
             Directory.Move(hgFolder, hgPath);
         }
-        await CleanupRepoFolder(repoPath);
-        SetPermissionsRecursively(dir);
+        await CleanupRepoFolder(tempRepoPath);
+        SetPermissionsRecursively(tempRepo);
+        // Now we're ready to move the new repo into place, replacing the old one
+        var realRepoPath = PrefixRepoFilePath(code);
+        await DeleteRepo(code); // Do this as late as possible
+        Directory.Move(tempRepoPath, realRepoPath);
```

Comment:
CoPilot told me that this essentially does a copy and delete under the hood, which would lose the permissions you assigned earlier. But it seems it was wrong. Maybe we can add this comment, because this sort of thing is not obvious to me.
```suggestion
        // Essentially just a rename, so permissions are preserved: https://github.com/dotnet/runtime/blob/5111fdc0dc464f01647d6b6078342f451bf3a499/src/libraries/System.Private.CoreLib/src/System/IO/FileSystem.Unix.cs#L451
        Directory.Move(tempRepoPath, realRepoPath);
```

Reply:
I don't know what it does on Windows, but on Linux this *should* be using the `mv` command (or rather, the Linux kernel API that `mv` calls), which will do an atomic move preserving permissions, if the dotnet programmers were halfway knowledgeable about Linux (and it seems they were). Copilot was probably trained on Windows code as there's a lot more of it than Linux code. Yet another example of why one shouldn't rely on generative AI to get facts right. :-)

Reply:
Done in commit b2f6ef07.